java-checks/src/main/java/org/sonar/java/checks/AssertTrueInsteadOfDedicatedAssertCheck.java

```diff
@@ -0,0 +1,205 @@
+    EQUALS("Equals", "An equals check"),
+    NOT_EQUALS("NotEquals", "An equals check");
+
+    public final String methodName;
+    public final String actionDescription;
+
+    Assertion(String namePostfix, String actionDescription) {
+      methodName = "assert" + namePostfix;
+      this.actionDescription = actionDescription;
+    }
+  }
+
+  @Override
+  protected MethodMatchers getMethodInvocationMatchers() {
+    return MethodMatchers.create().ofTypes(ASSERTION_CLASSES).names(ASSERT_METHOD_NAMES).withAnyParameters().build();
+  }
+
+  @Override
+  protected void onMethodInvocationFound(MethodInvocationTree mit) {
+    if (!hasSemantic()) {
+      return;
+    }
+
+    Arguments arguments = mit.arguments();
+
+    ExpressionTree argumentExpression;
+    if (hasBooleanArgumentAtPosition(arguments, 0)) {
+      argumentExpression = arguments.get(0);
+    } else if (hasBooleanArgumentAtPosition(arguments, arguments.size() - 1)) {
+      argumentExpression = arguments.get(arguments.size() - 1);
+    } else {
+      // We encountered a JUnit5 assert[True|False] method that accepts a BooleanSupplier - not supported.
+      return;
+    }
+
+    Optional<Assertion> replacementAssertionOpt = getReplacementAssertion(argumentExpression);
+
+    if (replacementAssertionOpt.isPresent()) {
+      IdentifierTree problematicAssertionCallIdentifier = ExpressionUtils.methodName(mit);
+      if (problematicAssertionCallIdentifier.name().equals("assertFalse")) {
+        replacementAssertionOpt = complement(replacementAssertionOpt.get());
+
+        if (!replacementAssertionOpt.isPresent()) {
+          return;
+        }
+      }
+
+      Assertion replacementAssertion = replacementAssertionOpt.get();
+
+      List<JavaFileScannerContext.Location> secondaryLocation = Collections.singletonList(new JavaFileScannerContext.Location(
+        String.format("%s is performed here, which is better expressed with %s.",
+          replacementAssertion.actionDescription, replacementAssertion.methodName),
+        argumentExpression));
+      String message = String.format("Use %s instead.", replacementAssertion.methodName);
+
+      reportIssue(problematicAssertionCallIdentifier, message, secondaryLocation, null);
+    }
+  }
+
+  private static boolean hasBooleanArgumentAtPosition(Arguments arguments, int index) {
+    return arguments.size() > index && arguments.get(index).symbolType().isPrimitive(Type.Primitives.BOOLEAN);
+  }
+
+  /**
+   * Returns the assertX method that should be used instead of assertTrue, if applicable.
+   *
+   * @param argumentExpression the boolean expression passed to assertTrue
+   * @return the assertion method to be used instead of assertTrue, or {@code null} if no better assertion method was determined
+   */
+  private static Optional<Assertion> getReplacementAssertion(@Nullable ExpressionTree argumentExpression) {
+    if (argumentExpression == null) {
+      return Optional.empty();
+    }
```

Comment:
`argumentExpression` can not be `null`, if we pass a value to `getReplacementAssertion` coming from methods not annotated by `@Nullable` or `@CheckForNull`. It's enough to ensure that it's never null.